torch_xla/csrc/runtime/cache.h

```diff
@@ -106,6 +122,116 @@ class Cache {
   ElementMap element_map_;
 };
 
+// A persistent cache which serializes values to disk. This wraps a Cache
+// instance, so values will only be read from disk once and subsequent reads
+// will go through the wrapped Cache.
+template <typename K, typename T, typename H = std::hash<K>,
+          typename E = std::equal_to<K>>
+class PersistentCache : public AbstractCache<K, T, H, E> {
+ public:
+  using TypePtr = std::shared_ptr<T>;
+
+  explicit PersistentCache(int kMaxMemoryCacheSize, std::string cache_dir,
+                           bool readonly,
+                           std::function<std::string(TypePtr&)> serialize,
+                           std::function<TypePtr(std::string&)> deserialize)
+      : memory_cache_(kMaxMemoryCacheSize),
+        cache_dir_(cache_dir),
+        readonly_(readonly),
+        serialize_(serialize),
+        deserialize_(deserialize) {
+    std::filesystem::create_directories(cache_dir);
+  }
+
+  // Add the value to the persistent cache. This only writes to disk if no
+  // existing value is tracked to avoid unnecessary serialization overhead.
+  // The value will also be tracked in memory, so subsequent Get calls will not
+  // incur deserialization.
+  // If the cache is readonly, nothing is written to disk.
+  TypePtr Add(K key, TypePtr obj) override {
+    std::lock_guard<std::mutex> slock(lock_);
+    std::string path = GetPath(key);
+    if (!Exists(path) && !readonly_) {
+      std::ofstream out(path, std::ios::binary);
+      out << serialize_(obj);
+    }
+    return memory_cache_.Add(key, obj);
+  }
+
+  // Get the TypePtr associated with the key. This method will first check
+  // if the key is tracked in memory, and if not it will check for a persisted
+  // version on disk.
+  TypePtr Get(const K& key) override {
+    std::lock_guard<std::mutex> slock(lock_);
+    TypePtr mem = memory_cache_.Get(key);
+    if (mem) {
+      return mem;
+    }
+
+    std::string path = GetPath(key);
+    if (!Exists(path)) {
+      TORCH_LAZY_COUNTER("PersistentCacheMiss", 1);
+      return nullptr;
+    }
+    std::stringstream ss;
+    std::ifstream in(path, std::ios::binary);
+    ss << in.rdbuf();
+    std::string serialization = ss.str();
+
+    TypePtr val = deserialize_(serialization);
+    if (!val) {
+      TORCH_LAZY_COUNTER("PersistentCacheDeserializeFailure", 1);
+      // Remove the serialized value from disk to allow a new value to be stored
+      _erase(key);
+      return nullptr;
+    }
+    TORCH_LAZY_COUNTER("PersistentCacheHit", 1);
+    // Make sure the memory_cache_ tracks the value to prevent multiple loads
+    return memory_cache_.Add(key, val);
+  }
+
+  void Clear() override {
+    std::lock_guard<std::mutex> slock(lock_);
+    memory_cache_.Clear();
+    // Delete and recreate the cache directory on disk.
+    if (!readonly_) {
```

Comment:
Maybe instead of conditioning on `readonly_` here, introduce an optional arg for `Clear` to clear the persistent cache, if `cache_dir` is provided.

Reply:
I think we should respect the `readonly_` setting here - the reason for it is in shared mounts, we only want the writer to modify the objects on disk.